### ISSUE TITLE
onMove receives inconsistent parentId and index

### ISSUE THREAD

**Issue by davidnaas** [brimdata/react-arborist#247]
**Inconsistent index**
Cursor would imply new index should be 1 but is reported as 2
![Kapture 2024-04-23 at 12 23 49](https://github.com/brimdata/react-arborist/assets/5172136/47813302-ce56-43c8-b101-edf8ad2cb4dd)

**Inconsistent parentId**
Cursor and `node.willReceiveDrop` indicates that a drop will **NOT** happen, yet `parentId` is that of the `Box` rather than `Container`

This behaviour only happens after first moving `Slider` into a `Box`, then out again. Somehow that `Slider` continues to receive the `Box` as the parent when being moved. It should also be noted that those boxes are internal nodes in all cases, i.e. `children` is always an array
![Kapture 2024-04-23 at 12 21 57](https://github.com/brimdata/react-arborist/assets/5172136/87541acf-ab94-4da4-b610-a484f2c779be)

**Comment by RAFA3L** [comment]
Same here, moving a node from a nested node to another one get an incorrect index, then if you move the node inside the same level the index is correct.

**Comment by surajair** [comment]
Yes. Same here. onMove is giving incorrect position. at root level, if there are two nodes, it am getting position as 2 whereas it should have been 1.

**Comment by RAFA3L** [comment]
Possible solution? instead of get the index from the args in onMove, yo can get the proper one from the tree state like this, where tree is the ref to the component.

tree.state.dnd.index;

**Comment by RAFA3L** [comment]
I discovered that the problem in my case is when the node next position is above the old one, even to another level but always dragging it upward, then the index jump one step more.

![2024-08-14_23-17-07](https://github.com/user-attachments/assets/1311bd8e-7372-4902-b11b-7e97ed0a4ead)

**Comment by surajair** [comment]
> Possible solution? instead of get the index from the args in onMove, yo can get the proper one from the tree state like this, where tree is the ref to the component.
> 
> 
> 
> tree.state.dnd.index;

Will try this

**Comment by RAFA3L** [comment]
@surajair the index in the state has the same problem, dragging the node upward the index is increased by 1, you can calculate the direction getting the old rowIndex of the node from the tree instance

**Comment by furkanalptokac** [comment]
Same here. When I drag a node above or below itself, index value changes. According to the example, index is 0 on above but index is 1 on below. It moves under next element when index is 1 with a simple dnd logic. That case is completely blocking me.
Uploading Screen Recording 2024-12-24 181332.mp4…

**Comment by furkanalptokac** [comment]
Same here. When I drag a node above or below itself, index value changes. According to the example, index is 0 on above but index is 1 on below. It moves under next element when index is 1 with a simple dnd logic. That case is completely blocking me.
https://github.com/user-attachments/assets/921fda75-1678-45df-bcb5-3d9be96c8c36

**Comment by furkanalptokac** [comment]
Same here. When I drag a node above or below itself, index value changes. According to the example, index is 0 on above but index is 1 on below. It moves under next element when index is 1 with a simple dnd logic. That case is completely blocking me.
![Screen Recording 2024-12-24 181332](https://github.com/user-attachments/assets/6ac4a2d3-09fe-4436-a208-bdc97c41229a)

**Comment by jameskerr** [comment]
Hey everyone here. I see two issues here.

1. Dragging a node between two empty internal nodes doesn't work.
2. When you drag the first child node "under itself" the index is 1.

The first issue is a bug and I will need to fix. 

The second issue might be a bug, might not. You may need to change your "onMove" implimentation. Here is how I have written my onMove function in the past.

```ts
  const onMove = (args: {
    dragIds: string[];
    parentId: null | string;
    index: number;
  }) => {
    for (const id of args.dragIds) {
      tree.move({ id, parentId: args.parentId, index: args.index });
    }
    setData(tree.data);
  };
```

And here is the tree.move function.

```ts
  move(args: { id: string; parentId: string | null; index: number }) {
    const src = this.find(args.id);
    const parent = args.parentId ? this.find(args.parentId) : this.root;
    if (!src || !parent) return;
    parent.addChild(src.data, args.index);
    src.drop();
  }
```

And here is the `addChild` function.

```ts
  addChild(data: T, index: number) {
    const node = createNode(data, this);
    this.children = this.children ?? [];
    this.children.splice(index, 0, node);
    this.data.children = this.data.children ?? [];
    this.data.children.splice(index, 0, data);
  }
```

And here is the drop function.

```ts
  drop() {
    if (this.hasParent()) this.parent.removeChild(this.childIndex);
  }
```

All this code is here in the repo https://github.com/brimdata/react-arborist/blob/aba9953a40a84d8491bcb9481fed52eb5f041ca3/modules/react-arborist/src/hooks/use-simple-tree.ts#L28

The algorithm works like this....

![Image](https://github.com/user-attachments/assets/274737d7-e9f4-4ac0-8ddf-d667caecace0)

**Comment by jameskerr** [comment]
@davidnaas What if you move your cursor further left before dropping?
